Chap_API_Storage.tex

```diff
@@ -0,0 +1,225 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Chapter: Storage support
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\chapter{Storage Support Definitions}
+\label{chap:api_storage}
+
+Distributed and parallel computing systems are increasingly embracing storage hierarchies to meet the diverse data management needs of applications and other systems software in a cost-effective manner.
+These hierarchies provide access to a number of distinct storage layers, with each potentially composed of different storage hardware (e.g., HDD, SSD, tape, PMEM), deployed at different locations (e.g., on-node, on-switch, on-site, WAN), and designed using different storage paradigms (e.g., file-based, object-based).
+Each of these systems offers unique performance and usage characteristics that storage system users should carefully consider to ensure the most efficient use of storage resources.
+
+PMIx enables users to better understand storage hierarchies by defining attributes that formalize storage system characteristics, state, and other parameters.
+These attributes can be queried by applications, I/O libraries and middleware, and workflow systems to discover available storage resources and to inform on which resources are most suitable for different I/O workload requirements.
+
+%%%%%%%%%%%
+\section{Storage support constants}
+
+\declarestruct{pmix_storage_type_t}
+
+The \refstruct{pmix_storage_type_t} structure is an enumerated type for specifying the type of storage system being used.
+
+\begin{constantdesc}
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_LUSTRE}
```

Comment:
Do these types need to be coupled with version information?  I'm not sure how much it matters from an application perspective what version of NFS, GPFS, or Lustre is being used.  But I suspect it matters quite a bit for the S3 and GCP types, and for the "younger" types like DAOS.

Reply:
That's a really good point. Even for things like Lustre, I'm sure it's useful for I/O middleware and other things to know what version is being used to optimize usage. So, something like this for the PMIX_STORAGE_TYPE definition (along the lines of the other thread above related to this attribute):

```
typedef struct pmix_storage_type
{
   char *type;
   char *version;
} pmix_storage_type_t;
```

Reply:
The changes that were made went the opposite direction of what I envisioned.

The removal of the actual constant values (e.g., "lustre") can cause interoperability issues between implementations and/or conversion steps between clients and servers, if the actual values varies for PMIX_STORAGE_TYPE_LUSTRE.

My initial problem was with the enumerated types for these constant values, not the fact that we do recognize some constant values as common/standardized.

Given that these are string types, we could just say that the following values are standardized: "lustre", "tmpfs", "gpfs", ... without providing the defined identifiers for them (i.e., not having #define PMIX_STORAGE_TYPE_GPFS "gpfs" in pmix.h).